File: back_end/schemas/commerce.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from schemas.product import ProductRead


class UserRead(BaseModel):
    id: int
    email: str
    name: str | None = None
    created_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class AuthRequest(BaseModel):
    email: str
    password: str = Field(..., min_length=6)
    name: str | None = None
    guest_id: str | None = None


class AuthResponse(BaseModel):
    user: UserRead
    session_token: str


class CartItemRead(BaseModel):
    product: ProductRead
    quantity: int
    line_total: float


class CartRead(BaseModel):
    items: list[CartItemRead]
    subtotal: float
    count: int


class CartItemRequest(BaseModel):
    product_id: int
    quantity: int = Field(default=1, ge=1, le=99)
    guest_id: str | None = None


class CartQuantityRequest(BaseModel):
    quantity: int = Field(..., ge=0, le=99)
    guest_id: str | None = None


class GuestRequest(BaseModel):
    guest_id: str | None = None


class CheckoutRequest(BaseModel):
    email: str
    full_name: str = Field(..., min_length=1)
    address: str = Field(..., min_length=1)
    city: str = Field(..., min_length=1)
    country: str = Field(..., min_length=1)
    phone: str | None = None
    guest_id: str | None = None


class OrderItemRead(BaseModel):
    product_id: int
    product_name: str
    unit_price: float
    quantity: int

    model_config = ConfigDict(from_attributes=True)


class OrderRead(BaseModel):
    id: int
    email: str
    full_name: str
    address: str
    city: str
    country: str
    phone: str | None = None
    status: str
    subtotal: float
    created_at: datetime | None = None
    items: list[OrderItemRead]

    model_config = ConfigDict(from_attributes=True)


class ShopAssistantRequest(BaseModel):
    message: str = Field(..., min_length=1)
    product_id: int | None = None


class CartDraft(BaseModel):
    product_id: int
    quantity: int
    rationale: str


class ShopAssistantResponse(BaseModel):
    message: str
    cart_draft: CartDraft | None = None

File: front_end/types/commerce.ts
import type { Product } from "@/types/product";

export type User = {
  id: number;
  email: string;
  name?: string | null;
  is_superadmin: boolean;
  is_active: boolean;
  created_at?: string | null;
};

export type CartItem = {
  product: Product;
  quantity: number;
  line_total: number;
};

export type Cart = {
  items: CartItem[];
  subtotal: number;
  count: number;
};

export type Order = {
  id: number;
  email: string;
  full_name: string;
  address: string;
  city: string;
  country: string;
  phone?: string | null;
  status: string;
  subtotal: number;
  created_at?: string | null;
  items: {
    product_id: number;
    product_name: string;
    unit_price: number;
    quantity: number;
  }[];
};

export type CartDraft = {
  product_id: number;
  quantity: number;
  rationale: string;
};

export type AdminOverview = {
  product_count: number;
  active_product_count: number;
  user_count: number;
  active_user_count: number;
  order_count: number;
  revenue_subtotal: number;
};

export type AdminUserList = {
  items: User[];
  total: number;
  page: number;
  page_size: number;
};

export type AdminOrderList = {
  items: Order[];
  total: number;
  page: number;
  page_size: number;
};
